cik_benchmark/baselines/direct_prompt.py

```diff
@@ -70,31 +71,71 @@ def dict_to_obj(data):
         return data
 
 
+import re
+from lmformatenforcer import JsonSchemaParser, RegexParser
+from lmformatenforcer.integrations.transformers import (
+    build_transformers_prefix_allowed_tokens_fn,
+)
+
+
 @torch.inference_mode()
 def huggingface_instruct_model_client(
-    llm, tokenizer, model, messages, n=1, max_tokens=10000, temperature=1.0, **kwargs
+    llm,
+    tokenizer,
+    model,
+    messages,
+    n=1,
+    max_tokens=10000,
+    temperature=1.0,
+    constrained_decoding=True,
+    future_timestamps=None,
+    **kwargs,
 ):
-    text = tokenizer.apply_chat_template(
-        messages, tokenize=False, add_generation_prompt=True
-    )
+    if constrained_decoding:
+        assert (
+            future_timestamps is not None
+        ), "Future timestamps must be provided for constrained decoding"
 
-    model_inputs = tokenizer([text], return_tensors="pt").to(llm.device)
-    batch = {k: v.repeat(n, 1) for k, v in model_inputs.items()}
-    num_input_ids = batch["input_ids"].shape[1]
+    def constrained_decoding_regex(required_timestamps):
+        """
+        Generates a regular expression to force the model output
+        to satisfy the required format and provide values for
+        all required timestamps
 
-    generated_ids = llm.generate(
-        **batch, temperature=temperature, max_new_tokens=max_tokens
+        """
+        timestamp_regex = "".join(
+            [
+                r"\(\s*{}\s*,\s*[-+]?\d+(\.\d+)?\)\n".format(re.escape(ts))
+                for ts in required_timestamps
+            ]
+        )
+        return r"<forecast>\n{}<\/forecast>".format(timestamp_regex)
+
+    # Make generation pipeline
+    pipe = pipeline(
+        task="text-generation",
+        model=llm,
+        tokenizer=tokenizer,
+        device_map="auto",
     )
 
-    batch_responses = tokenizer.batch_decode(
-        generated_ids[:, num_input_ids:], skip_special_tokens=True
+    # Build a regex parser with the generated regex
+    parser = RegexParser(constrained_decoding_regex(future_timestamps))
+    prefix_function = build_transformers_prefix_allowed_tokens_fn(
+        pipe.tokenizer, parser
     )
 
     # Now extract the assistant's reply
     choices = []
-    for response_text in batch_responses:
+    for response in pipe(
+        [messages] * n,
+        max_length=max_tokens,
+        temperature=temperature,
+        prefix_allowed_tokens_fn=prefix_function,
+        batch_size=n,
+    ):
         # Create a message object
-        message = SimpleNamespace(content=response_text)
+        message = SimpleNamespace(content=response[0]["generated_text"][-1]["content"])
```

Comment:
Would this not break previous code that does normal decoding?

Reply:
No, this is how pipelines output, not related to decoding.